File: NasClientDesktop/Services/Format.cs
// NasClientDesktop/Services/Format.cs
// 展示与校验辅助。

using System;
using System.Collections.Generic;

namespace NasClientDesktop.Services;

public enum PreviewKind { None, Image, Video, Audio, Text }

public static class Format
{
    /// <summary>1234567 → "1.2 MB"。</summary>
    public static string Size(long bytes)
    {
        if (bytes < 1024) return $"{bytes} B";
        string[] units = ["KB", "MB", "GB", "TB"];
        double value = bytes;
        var unit = "B";
        foreach (var u in units)
        {
            value /= 1024;
            unit = u;
            if (value < 1024) break;
        }
        return value >= 100 ? $"{value:0} {unit}" : $"{value:0.0} {unit}";
    }

    /// <summary>UTC 时间 → 本地 "2026-06-11 09:30"。</summary>
    public static string Time(DateTime utc)
    {
        if (utc == default) return "—";
        var local = utc.Kind == DateTimeKind.Utc ? utc.ToLocalTime()
            : DateTime.SpecifyKind(utc, DateTimeKind.Utc).ToLocalTime();
        return local.ToString("yyyy-MM-dd HH:mm");
    }

    /// <summary>把路径拆成面包屑段。</summary>
    public static List<string> SplitPath(string path)
    {
        var result = new List<string>();
        foreach (var seg in path.Split('/'))
            if (seg.Length > 0) result.Add(seg);
        return result;
    }

    /// <summary>拼接相对路径，去除多余分隔符。</summary>
    public static string JoinPath(params string[] parts)
    {
        var segs = new List<string>();
        foreach (var p in parts)
            foreach (var s in p.Split('/'))
                if (s.Length > 0) segs.Add(s);
        return string.Join('/', segs);
    }

    /// <summary>拼接本地 OS 路径（用平台分隔符，经 NasLib）。用于下载落地路径。</summary>
    public static string JoinPathLocal(string dir, string name) => Lukas.Std.Io.Path.Combine(dir, name);

    /// <summary>客户端预检：目录/文件名里不允许的输入。返回错误信息或 null。</summary>
    public static string? ValidateNameSegment(string name)
    {
        var trimmed = name.Trim();
        if (trimmed.Length == 0) return "名称不能为空";
        if (trimmed is "." or "..") return "名称不能是 \".\" 或 \"..\"";
        foreach (var ch in trimmed)
        {
            if (ch is '\\' or '/' or ':' or '*' or '?' or '"' or '<' or '>' or '|')
                return "名称含有不允许的字符（\\ / : * ? \" < > |）";
            if (ch < 0x20) return "名称含有控制字符";
        }
        return trimmed.Length > 255 ? "名称过长（最多 255 字符）" : null;
    }

    // ---------------------------------------------------------------- 文件类型识别

    private static readonly HashSet<string> ExtImage = new(StringComparer.OrdinalIgnoreCase)
        { "jpg", "jpeg", "png", "gif", "webp", "bmp", "svg", "ico", "avif" };
    private static readonly HashSet<string> ExtVideo = new(StringComparer.OrdinalIgnoreCase)
        { "mp4", "webm", "ogv", "mov", "mkv" };
    private static readonly HashSet<string> ExtAudio = new(StringComparer.OrdinalIgnoreCase)
        { "mp3", "wav", "ogg", "oga", "flac", "aac", "m4a", "opus" };
    private static readonly HashSet<string> ExtText = new(StringComparer.OrdinalIgnoreCase)
    {
        "txt", "md", "markdown", "log", "json", "xml", "csv", "tsv", "yaml", "yml", "ini", "conf",
        "html", "htm", "css", "js", "jsx", "ts", "tsx", "c", "h", "cpp", "hpp", "cs", "java", "py",
        "rb", "go", "rs", "php", "sh", "bash", "sql", "toml", "env", "gitignore", "dockerfile",
    };

    private static string ExtOf(string name)
    {
        var dot = name.LastIndexOf('.');
        return dot >= 0 ? name[(dot + 1)..] : "";
    }

    /// <summary>Avalonia 内置位图可直接解码的图片格式（其余如 svg/ico/avif 走系统默认程序）。</summary>
    private static readonly HashSet<string> ExtBitmapNative = new(StringComparer.OrdinalIgnoreCase)
        { "jpg", "jpeg", "png", "gif", "webp", "bmp" };

    public static bool IsBitmapNative(string name) => ExtBitmapNative.Contains(ExtOf(name));

    public static PreviewKind PreviewKindOf(string name)
    {
        var ext = ExtOf(name);
        if (ExtImage.Contains(ext)) return PreviewKind.Image;
        if (ExtVideo.Contains(ext)) return PreviewKind.Video;
        if (ExtAudio.Contains(ext)) return PreviewKind.Audio;
        return ExtText.Contains(ext) ? PreviewKind.Text : PreviewKind.None;
    }
}

File: nas-client/src/lib/format.ts
// src/lib/format.ts

/** 1234567 → "1.2 MB"（文件系统素材，配合等宽字体展示）。 */
export function formatSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  const units = ['KB', 'MB', 'GB', 'TB'] as const
  let value = bytes
  let unit: string = 'B'
  for (const u of units) {
    value /= 1024
    unit = u
    if (value < 1024) break
  }
  return `${value >= 100 ? value.toFixed(0) : value.toFixed(1)} ${unit}`
}

/** ISO 时间 → 本地 "2026-06-11 09:30"。 */
export function formatTime(iso: string): string {
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) return '—'
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}`
}

/** 把路径拆成面包屑段：'docs/2026' → ['docs', '2026']。 */
export function splitPath(path: string): string[] {
  return path.split('/').filter((s) => s.length > 0)
}

/** 拼接相对路径，自动去除多余分隔符。 */
export function joinPath(...parts: string[]): string {
  return parts
    .flatMap((p) => p.split('/'))
    .filter((s) => s.length > 0)
    .join('/')
}

/** 客户端预检：目录/文件名里不允许的输入（与后端 SafeName 同向，最终以后端校验为准）。 */
export function validateNameSegment(name: string): string | null {
  const trimmed = name.trim()
  if (trimmed.length === 0) return '名称不能为空'
  if (trimmed === '.' || trimmed === '..') return '名称不能是 "." 或 ".."'
  if (/[\\/:*?"<>|]/.test(trimmed)) return '名称含有不允许的字符（\\ / : * ? " < > |）'
  for (const ch of trimmed) {
    const code = ch.codePointAt(0) ?? 0
    if (code < 0x20) return '名称含有控制字符'
  }
  if (trimmed.length > 255) return '名称过长（最多 255 字符）'
  return null
}

// ---------------------------------------------------------------- 文件类型识别
export type PreviewKind = 'image' | 'video' | 'audio' | 'text' | 'none'

const EXT_IMAGE = new Set(['jpg', 'jpeg', 'png', 'gif', 'webp', 'bmp', 'svg', 'ico', 'avif'])
const EXT_VIDEO = new Set(['mp4', 'webm', 'ogv', 'mov', 'mkv'])
const EXT_AUDIO = new Set(['mp3', 'wav', 'ogg', 'oga', 'flac', 'aac', 'm4a', 'opus'])
const EXT_TEXT = new Set([
  'txt', 'md', 'markdown', 'log', 'json', 'xml', 'csv', 'tsv', 'yaml', 'yml', 'ini', 'conf',
  'html', 'htm', 'css', 'js', 'jsx', 'ts', 'tsx', 'c', 'h', 'cpp', 'hpp', 'cs', 'java', 'py',
  'rb', 'go', 'rs', 'php', 'sh', 'bash', 'sql', 'toml', 'env', 'gitignore', 'dockerfile',
])

function extOf(name: string): string {
  const dot = name.lastIndexOf('.')
  return dot >= 0 ? name.slice(dot + 1).toLowerCase() : ''
}

/** 根据文件名判断可用的在线预览方式。 */
export function previewKindOf(name: string): PreviewKind {
  const ext = extOf(name)
  if (EXT_IMAGE.has(ext)) return 'image'
  if (EXT_VIDEO.has(ext)) return 'video'
  if (EXT_AUDIO.has(ext)) return 'audio'
  if (EXT_TEXT.has(ext)) return 'text'
  return 'none'
}
